clang/docs/LanguageExtensions.rst

```diff
@@ -3774,6 +3774,74 @@ type-generic alternative to the ``__builtin_clz{,l,ll}`` (respectively
 ``__builtin_ctz{,l,ll}``) builtins, with support for other integer types, such
 as ``unsigned __int128`` and C23 ``unsigned _BitInt(N)``.
 
+``__builtin_counted_by_ref``
+----------------------------
+
+``__builtin_counted_by_ref`` returns a pointer to the count field from the
+``counted_by`` attribute.
+
+The argument must be a pointer to a flexible array member. If the argument
+isn't a flexible array member or doesn't have the ``counted_by`` attribute, the
```

Comment:
Well, what I mean is that this is actually an error currently to add any other bounds attributes on a flexible array member, because other attributes don't make sense. For example, for counted_by_or_null, a flexible array member cannot be a null because it's not a pointer. And sized_by* is only allowed on pointers. Hence, these are naturally prevented right now, and if we extend the builtin to pointers we may want to emit an error for sized_by* because it has a different scale than counted_by*, but we can discuss that later when we actually support it.

Reply:
But is this the best place to talk about which attributes are valid on a flexible array member? IMHO, that comment should be closer to the documentation for those attributes; in a similar vein, the checks for those situations should be where the attributes are handled during parsing.

Reply:
I thought it would be relevant in the interest for this builtin's behavior for other bounds attributes. We would want to define the behavior for them in the future when this builtin gets exposed to pointers where these attributes are possible (which would be basically to decide whether we'd like to diagnose it or ignore them silently), but it's not currently exposed yet.

Reply:
Okay, do you have an example of what to add? I'm coming up with clunky ways of saying it. :smile: